### PR TITLE
Add Code Mining support

### PR DESCRIPTION
**Pull request by kthoms**
This PR adds support of JFace's new Code Mining API. 

It introduced a new bundle `org.eclipse.xtext.ui.codemining`, which adds this support. The bundle will only be runnable on an Eclipse 4.8+ target, but fail safely on an older target. Client DSLs will not fail when installed on Eclipse Oxygen or older when implementing the API, but will miss the function of course.

Implemented an example mining for the Domainmodel DSL.

### CODE REVIEW

File: org.eclipse.xtext.ui.codemining/src/org/eclipse/xtext/ui/codemining/AbstractXtextCodeMiningProvider.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+				}
+				codeMiningList.add(codeMiningObject);
+			}
+		};
+		
+		if (cancelIndicator.isCanceled()) {
+			throw new CancellationException();
+		}
+		// first mine for header annotations, then inline annotations
+		createLineHeaderCodeMinings(document, resource, cancelIndicator, acceptor);
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.codemining/src/org/eclipse/xtext/ui/codemining/AbstractXtextCodeMiningProvider.java]: I'm not yet convinced that we do need two abstract methods here.

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/.project [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/.project]: What is this?

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src-gen/org/eclipse/xtext/example/domainmodel/ui/AbstractDomainmodelUiModule.java [hunk 3]
```diff
@@ -208,6 +212,7 @@ public void configureIResourceDescriptionsPersisted(Binder binder) {
 	}
 	
 	// contributed by org.eclipse.xtext.xtext.generator.generator.GeneratorFragment2
+	@Override
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src-gen/org/eclipse/xtext/example/domainmodel/ui/AbstractDomainmodelUiModule.java]: Edited manually?

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src/org/eclipse/xtext/example/domainmodel/ui/codemining/DomainmodelCodeMiningProvider.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	@Inject
+	private DomainmodelGrammarAccess grammar;
+	
+	@Override
+	protected void createLineHeaderCodeMinings(IDocument document, XtextResource resource, CancelIndicator indicator, IAcceptor<ICodeMining> acceptor) 
+			throws BadLocationException{
+		int lineCount = document.getNumberOfLines();
+		for (int i = 0; i < lineCount; i++) {
+			String line = document.get(document.getLineOffset(i), document.getLineLength(i)).trim();
+			if(line.contains("op")) {
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src/org/eclipse/xtext/example/domainmodel/ui/codemining/DomainmodelCodeMiningProvider.java]: We should fix the example

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/.project (hunk 2 above)
Comment (kthoms) [reply org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/.project]: by accident created by project import action, will delete this

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src/org/eclipse/xtext/example/domainmodel/ui/codemining/DomainmodelCodeMiningProvider.java (hunk 4 above)
Comment (kthoms) [reply org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src/org/eclipse/xtext/example/domainmodel/ui/codemining/DomainmodelCodeMiningProvider.java]: in which regard?

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src-gen/org/eclipse/xtext/example/domainmodel/ui/AbstractDomainmodelUiModule.java (hunk 3 above)
Comment (kthoms) [reply org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src-gen/org/eclipse/xtext/example/domainmodel/ui/AbstractDomainmodelUiModule.java]: will revert all additions of override that were added

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src/org/eclipse/xtext/example/domainmodel/ui/codemining/DomainmodelCodeMiningProvider.java (hunk 4 above)
Comment (szarnekow) [reply org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src/org/eclipse/xtext/example/domainmodel/ui/codemining/DomainmodelCodeMiningProvider.java]: line.contains("op") is a rather weak constraint, e.g. Properties names 'propertyName' or 'top' won't work.

File: org.eclipse.xtext.ui.codemining/src/org/eclipse/xtext/ui/codemining/AbstractXtextCodeMiningProvider.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	 * Clients have to implement this method to provide code minings.
+	 * 
+	 * @param document The document
+	 * @param resource The resource for that document
+	 * @param indicator Cancelation indicator
+	 * @param acceptor Accepts created minings
+	 * @throws BadLocationException when line number doesn't exists
+	 */
+	protected abstract void createCodeMinings(IDocument document, XtextResource resource,
+			CancelIndicator indicator, IAcceptor<ICodeMining> acceptor) throws BadLocationException;
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.codemining/src/org/eclipse/xtext/ui/codemining/AbstractXtextCodeMiningProvider.java]: Signature should be `IAcceptor<? super ICodeMining>`

File: org.eclipse.xtext.ui.codemining/src/org/eclipse/xtext/ui/codemining/XtextCodeMiningReconcileStrategy.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+			}
+			return null;
+		}
+
+	}
+
+	private ISourceViewerExtension5 sourceViewer;
+
+	public XtextCodeMiningReconcileStrategy(ISourceViewerExtension5 sourceViewer) {
+			this.sourceViewer = sourceViewer;
```
Comment (szarnekow) [inline org.eclipse.xtext.ui.codemining/src/org/eclipse/xtext/ui/codemining/XtextCodeMiningReconcileStrategy.java]: Indentation

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/.project (hunk 2 above)
Comment (szarnekow) [reply org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/.project]: Looks like it is still here

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/plugin.xml [hunk 7]
```diff
@@ -649,4 +668,5 @@
 			</activeWhen>
 		</handler>
 	</extension>
+	<!-- manual additions end here -->
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/plugin.xml]: ?

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/plugin.xml [hunk 8]
```diff
@@ -603,8 +603,27 @@
...
 	</extension>
-	<!-- manual additions start here -->
 
+	<!-- code mining -->
+	<extension point="org.eclipse.ui.workbench.texteditor.codeMiningProviders">
+		<codeMiningProvider
+			class="org.eclipse.xtext.example.domainmodel.ui.DomainmodelExecutableExtensionFactory:org.eclipse.jface.text.codemining.ICodeMiningProvider"
+			id="org.eclipse.xtext.example.domainmodel.Domainmodel.CodeMiningProvider"
+			label="Domainmodel Codemining">
+			<enabledWhen>
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/plugin.xml]: @cdietrich What do you think? Should we enable the minings only on the XtextEditor or on all editors that use the same Document type

File: org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src/org/eclipse/xtext/example/domainmodel/ui/codemining/DomainmodelCodeMiningProvider.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+		List<Operation> allOperations = EcoreUtil2.eAllOfType(resource.getContents().get(0), Operation.class);
+		// get keyword for ')'
+		Keyword rightParenthesisKeyword_4 = grammar.getOperationAccess().getRightParenthesisKeyword_4();
+		for (Operation o : allOperations) {
+			//inline annotations only for methods with no return type
+			if (o.getType() != null) {
+				continue;
+			}
+			// get return type name from operation
+			JvmOperation inferredOp = (JvmOperation) jvmModelAssociations.getPrimaryJvmElement(o);
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui/src/org/eclipse/xtext/example/domainmodel/ui/codemining/DomainmodelCodeMiningProvider.java]: We need to check this against broken models, e.g. ops without a body

**Pull request merged**